gematria/datasets/convert_bhive_to_llvm_exegesis_input.cc

```diff
@@ -240,22 +215,47 @@ int main(int argc, char* argv[]) {
       return 3;
     }
 
-    auto proto = bhive_importer.BasicBlockProtoFromMachineCode(*bytes);
+    llvm::Expected<std::vector<gematria::DisassembledInstruction>>
+        instructions = gematria::DisassembleAllInstructions(
+            llvm_support->mc_disassembler(), llvm_support->mc_instr_info(),
+            llvm_support->mc_register_info(), llvm_support->mc_subtarget_info(),
+            *inst_printer, 0, *bytes);
 
     // Check for errors.
-    if (!proto.ok()) {
-      std::cerr << "Failed to disassemble block '" << hex << ": "
-                << proto.status() << "\n";
+    if (!instructions) {
+      std::cerr << "Failed to disassemble block '" << hex << "\n";
       continue;
     }
 
+    // Get used registers
+    std::map<unsigned, bool> used_registers;
+    for (const gematria::DisassembledInstruction& instruction : *instructions) {
+      for (unsigned operand_index = 0;
+           operand_index < instruction.mc_inst.getNumOperands();
+           ++operand_index) {
+        if (instruction.mc_inst.getOperand(operand_index).isReg()) {
+          unsigned register_number =
+              instruction.mc_inst.getOperand(operand_index).getReg();
+          if (register_number == 0) continue;
+          unsigned super_register_number = 0;
+          for (MCPhysReg testing :
+               reg_info.superregs_inclusive(register_number)) {
+            super_register_number = testing;
+          }
```

Comment:
This stops at the _last_ super-register. Is there a guarantee that it is also the largest super-register? If not, we should either recurse (whenever we find a proper super-register, step up and look for its own super-registers), or filter by register class and take only the 64-bit superreg, if any.

Though the general idea is probably safer and more general.

Reply:
From everything that I can tell, yes. It's not explicitly enumerated in the documentation for the iterator, but the iterator internally starts at the given register and then goes up one level at a time (from what I can tell). Would probably be good to open a patch on the LLVM side to document this.